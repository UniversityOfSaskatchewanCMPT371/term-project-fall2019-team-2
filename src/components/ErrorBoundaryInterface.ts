export interface ErrorBoundaryInterface {
  /**
   * The sentry Event ID
   */
  eventId?: string,
  /**
   * True if an error has occurred
   */
  hasError?: boolean,
  /**
   * The stored error
   */
  error: Error,
  /**
   * Any additional error metadata
   */
  errorInfo?: any
}
