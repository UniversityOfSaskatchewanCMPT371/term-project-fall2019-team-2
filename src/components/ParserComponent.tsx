import React from 'react';
import ParserInterface, {ParserState} from './ParserInterface';
import Column, {enumDrawType} from './Column';
import moment from 'moment';
import * as d3
  from 'd3';
import TimelineComponent
  from './TimelineComponent';
import Data
  from './Data';
import * as TimSort from 'timsort';
import 'bootstrap/dist/css/bootstrap.min.css';

/**
 * Purpose: react component responsible for receiving and parsing file data
 */
export default class ParserComponent extends React.Component<ParserInterface,
  ParserState> {
    private columnTypes?: Array<Column>;
    private childKey = 0;

    /**
     * Purpose: ParserComponent constructor
     * @param {ParserInterface} props: the prompt and fileType properties to
     * pass into the constructor
     */
    constructor(props: ParserInterface) {
      super(props);
      this.state = {
        prompt: props.prompt,
        fileType: props.fileType,
        data: [],
        showTimeline: false,
        formatString: '',
      };

      this.isValid = this.isValid.bind(this);
      this.sortData = this.sortData.bind(this);
      this.inferTypes = this.inferTypes.bind(this);
      this.parseCsv = this.parseCsv.bind(this);
      this.parse = this.parse.bind(this);
      this.inferTypes = this.inferTypes.bind(this);
    }

    /**
     * Waits until component mounts
     */
    componentDidMount(): void {
    }

    /**
     * Purpose: renders the HTML for this component
     * @return {string}: valid HTML
     */
    render() {
      const chart = this.state.showTimeline ?
            <div>
              <TimelineComponent key={this.childKey}
                data={new Data('path/to/file', this.state.data,
                    this.columnTypes)}/>
            </div> :
            <div/>;

      return (
        <div>
          <div className="row">
            <div className="col-6">
              <label>
                {this.props.prompt}
              </label>
              <input
                type="file"
                onChange={this.parse}
                accept={this.props.fileType.mimeName}/>
            </div>
            <div className="col-6">
              <select value={this.state.formatString}
                onChange={(e) => {
                  const val = e.target.value;
                  this.setState(() => {
                    return {
                      formatString: val,
                    };
                  });
                }}>
                <option selected value="">Open this select menu</option>
                <option value="YYYY-MM-DD">YYYY-MM-DD</option>
                <option value="MM-DD-YYYY">MM-DD-YYYY</option>
                <option value="DD-MM-YYYY">DD-MM-YYYY</option>
                <option value="DD-MMMM-YYYY">DD-MMMM-YYYY</option>
                <option value="MMMM-DD-YYYY">MMMM-DD-YYYY</option>
                <option value="DD-MM-YY">DD-MM-YY</option>
                <option value="MM-DD-YY">MM-DD-YY</option>
              </select>
            </div>
          </div>

          <div className="row">
            <div className="col-12">
              {chart}
            </div>
          </div>
        </div>
      );
    }

    /**
     * Purpose: checks if the passed in event contains a file upload, then
     * verifies that the file type and contents are valid
     * @precondition no other parser object exists
     * @postcondition a parser object is instantiated as the only parser object
     * @param {Object} upFile: takes in the file
     * @return {boolean}: a boolean indicating whether or not the file upload is
     * valid
     */
    isValid(upFile?: File): boolean {
      if (upFile !== undefined) {
        const typeOfFile = upFile.name.substr(upFile.name.length - 4);
        if (typeOfFile === '.csv') {
          return typeOfFile === '.csv';
        } else {
          throw new Error('Wrong file type was uploaded.');
        }
      }
      throw new Error('Wrong file type was uploaded.');
    }

    /**
     * Purpose: sorts the array of data
     * @precondition dates must contain year month and date,
     *    if data does not contain year in some
     *    dates but does in some it will sort lexicographically.
     *    a csv has been uploaded, and the data is stored in an array.
     * @postcondition the data stored in the array is sorted by some date column
     * @param {Array} data: the array of data to sort
     * @return {boolean}: array of sorted data
     */
    sortData(data: Array<object>): boolean {
      let doneTheWork = false;
      /* loop goes through each key and saves the 1 with a date in first row */
      if (data[0] !== undefined && data.length > 0) {
        for (const [key, value] of Object.entries(data[0])) {
          if (!doneTheWork) {
            const date1 = moment(String(value));
            if (!isNaN(Number(date1)) && isNaN(Number(value))) {
              doneTheWork = true;
              const formatString = this.state.formatString;

              const keyInt = `${key}_num`;

              TimSort.sort(data, function(a: any, b: any) {
                let val: any;
                if (!a.hasOwnProperty(keyInt)) {
                  val = moment(a[key], formatString);
                  if (val.isValid()) {
                    a[keyInt] = val.valueOf();
                  } else {
                    a[keyInt] = -1;
                  }
                }

                if (!b.hasOwnProperty(keyInt)) {
                  val = moment(b[key], formatString);
                  if (val.isValid()) {
                    b[keyInt] = val.valueOf();
                  } else {
                    b[keyInt] = -1;
                  }
                }
                return (a[keyInt] - b[keyInt]);
              });

              this.setState(() => {
                return {
                  data,
                };
              });
            }
          }
        }
      }
      if (doneTheWork) {
        return true;
      } else {
        throw new Error('The file uploaded has no dates.');
      }
    }

    /**
     * Purpose: to instantiate an empty list of objects
     * for tracking the kinds of data in a column
     * @param {number} fieldLength: the number of columns of data
     * @return {[CountTypes]}: a list of objects
     */
    createTypeCountingObjects(fieldLength: number): CountTypes[] {
      const typesForEachCol = [];
      // instantiate object for each column
      for (let i = 0; i < fieldLength; i++) {
        typesForEachCol.push(new CountTypes());
      }
      return typesForEachCol;
    }

    /**
     * Purpose: attempts to infer the types of the data in each of the columns
     * of the csv data
     * @precondition An array of sorted data exists for types to be inferred
     * from
     * @postcondition The array is transformed into an array of type Column,
     * and the default behavior for the data is inferred and set
     * @param {Array} data: the array of pre-sorted valid data
     * @return {Array}: a list of objects of type Column
     */
    inferTypes(data: Array<object>): Array<Column> | undefined {
      if (this.state.data.length > 0) {
        const listFields = Object.keys(this.state.data[0]);
        // instantiate objects to track the types of data
        const typesForEachCol =
            this.createTypeCountingObjects(listFields.length);
        // check half the values to find if the data is consistent
        [0, Math.floor(this.state.data.length / 2)].forEach((element) => {
          const row: object = this.state.data[element];
          // look at each field and categorize
          for (let i = 0; i < listFields.length; i++) {
            const curColTypes = typesForEachCol[i];
            try {
              // @ts-ignore
              const val = row[listFields[i]];
              if (typeof val === 'string') {
                const date = moment(val);
                const isValid = date.isValid();
                if (isValid) {
                  curColTypes['numDate'] += 1;
                } else {
                  throw val;
                }
              } else {
                throw val;
              }
            } catch {
              // @ts-ignore
              const type = typeof row[listFields[i]];
              if (type !== 'string' && type !== 'number') {
                curColTypes['numIncongruent'] += 1;
              }
              // logs all the types that are seen
              if (type === 'string') {
                curColTypes['numString'] += 1;
              } else {
                curColTypes['numNumber'] += 1;
              }
            }
          }
        });
        let indx = 0;
        const arrayOfColumns = new Array<Column>(listFields.length);
        typesForEachCol.forEach((element) => {
          // checks the most common type and uses that
          const mostCommonType = element.largest();
          let newCol: Column;
          if (mostCommonType === 'string') {
            // create a Column object with occurrence data
            newCol = new Column(mostCommonType,
                enumDrawType.occurrence, listFields[indx]);
            arrayOfColumns[indx] = newCol;
            indx++;
          } else if (mostCommonType === 'number') {
            // create a Column with interval, point or magnitude data
            newCol = new Column(mostCommonType,
                enumDrawType.any, listFields[indx]);
            arrayOfColumns[indx] = newCol;
            indx++;
          } else if (mostCommonType === 'date') {
            // create a Column with date data
            // eslint-disable-next-line max-len
            newCol = new Column(mostCommonType, enumDrawType.any, listFields[indx]);
            arrayOfColumns[indx] = newCol;
            indx++;
          }
        }
        );
        return arrayOfColumns;
      } else {
        throw new Error('data is empty');
      }
    }

    /**
     * Purpose: attempts to parse the file uploaded by the user.
     * @precondition The user uploads a file
     * @postcondition The data is read out of the csv file and put into
     * an array object
     * @param {Object} fileEvent: the event passed into this component
     */
    async parse(fileEvent: any) {
      this.setState(() => {
        return {
          showTimeline: false,
        };
      });

      const temp: File = fileEvent.target.files[0];
      try {
        if (this.props.fileType.mimeName === '.csv' +
            ',text/csv' && this.isValid(temp)) {
          await this.parseCsv(fileEvent);
        }
      } catch (e) {
        alert('Wrong file type was uploaded.');
        console.log('Wrong file was uploaded.');
      }

      this.setState(() => {
        return {
          showTimeline: true,
        };
      });
      this.childKey++;
    }

    /**
     * Purpose: to parse a csv file uploaded by the user
     * @precondtion The user has uploaded a csv file
     * @postcondtion The data from the file is stored in a global array
     * @param {Object} fileEvent: the event passed into this component
     */
    async parseCsv(fileEvent: any) {
      const csvFile = fileEvent.target.files[0];

      // for testing
      this.props.onChange(fileEvent.target.files[0]);

      const fileReader = new FileReader();

      return new Promise((resolver, agent) => {
        const handleFileRead = () => {
          if (typeof fileReader.result === 'string') {
            const content = d3.csvParse(fileReader.result,
                function(d: any, i: number): any {
                  // autoType the row
                  d = d3.autoType(d);
                  // must add an index to the row to be used by the Timeline
                  d['index'] = i;

                  return d;
                });

            // set state of the parser component
            this.setState((state) => {
              return {
                prompt: this.state.prompt,
                fileType: this.state.fileType,
                data: content,
              };
            });
            try {
              this.columnTypes = this.inferTypes(this.state.data);
              this.sortData(this.state.data);
            } catch (e) {
              alert('data is EMPTY');
              console.log('data is empty');
            }
          }
          resolver(true);
        };

        fileReader.onloadend = handleFileRead;
        fileReader.readAsText(csvFile);
      });
    }
}

/**
 * Purpose: assist in counting types of data in a single Column
 */
export class CountTypes {
  public numNumber: number;
  public numIncongruent: number;
  public numString: number;
  public numDate: number;
  /**
   * create an empty component to track data
   */
  constructor() {
    this.numNumber = 0;
    this.numString = 0;
    this.numDate = 0;
    this.numIncongruent = 0;
  }

  /**
   * finds the largest element of the fields
   * @return {string}: a string representing the largest field
   */
  largest(): string {
    if (this.numDate >= this.numIncongruent && this.numDate >= this.numNumber) {
      if (this.numDate >= this.numString) {
        return 'date';
      } else {
        return 'string';
      }
    }
    if (this.numString >= this.numNumber &&
      this.numString >= this.numIncongruent) {
      return 'string';
    }
    if (this.numNumber >= this.numIncongruent) {
      return 'number';
    } else {
      return 'incongruent';
    }
  }
}
