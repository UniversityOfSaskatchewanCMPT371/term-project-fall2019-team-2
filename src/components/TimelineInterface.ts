import Data from './Data';

interface TimelineInterface {
    data: Data;
}
export default TimelineInterface;

export interface TimelineState {
    data: Data;
    // dimensions
    width: number;
    height: number;
    marginTop: number;
    marginBottom: number;
    marginLeft: number;
    marginRight: number;

    toggleTimeline: number;
    // Prompt for toggle button
    togglePrompt: string;

    xColumn: string;
    xColumn2: string;
    yColumn: string;

    loading: boolean;
    view: string;
}
