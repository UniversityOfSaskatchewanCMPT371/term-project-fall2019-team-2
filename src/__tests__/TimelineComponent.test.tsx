import React, {ReactDOM} from 'react';
import {mount, shallow} from 'enzyme';
import ParserComponent from '../components/TimelineComponent';
import {TimelineState} from '../components/TimelineInterface';
import sinon from 'sinon';

describe('<TimelineComponent /> renders correctly', () => {
  // todo: simple render test
});

// To be used by developers
describe('<TimelineComponent /> Unit Tests', () => {
  const noTest = 'No tests written for TimelineComponent ';

  describe('constructor()', () => {
    it('No unit tests written', () => {
      throw new Error(noTest + 'constructor()');
    });
  });

  describe('componentDidMount()', () => {
    it('No unit tests written', () => {
      throw new Error(noTest + 'componentDidMount()');
    });
  });

  describe('drawTimeline()', () => {
    it('No unit tests written', () => {
      throw new Error(noTest + 'drawTimeline()');
    });
  });

  describe('ttOver()', () => {
    it('No unit tests written', () => {
      throw new Error(noTest + 'ttOver()');
    });
  });

  describe('ttUpdatePos()', () => {
    it('No unit tests written', () => {
      throw new Error(noTest + 'ttUpdatePos()');
    });
  });

  describe('ttMove()', () => {
    it('No unit tests written', () => {
      throw new Error(noTest + 'ttMove()');
    });
  });

  describe('ttLeave()', () => {
    it('No unit tests written', () => {
      throw new Error(noTest + 'ttLeave()');
    });
  });

  describe('updateChart()', () => {
    it('No unit tests written', () => {
      throw new Error(noTest + 'updateChart()');
    });
  });

  describe('updateBars()', () => {
    it('No unit tests written', () => {
      throw new Error(noTest + 'updateBars()');
    });
  });

  describe('moveChart()', () => {
    it('No unit tests written', () => {
      throw new Error(noTest + 'moveChart()');
    });
  });

  describe('dragStarted()', () => {
    it('No unit tests written', () => {
      throw new Error(noTest + 'dragStarted()');
    });
  });

  describe('dragged()', () => {
    it('No unit tests written', () => {
      throw new Error(noTest + 'dragged()');
    });
  });

  describe('dragEnded()', () => {
    it('No unit tests written', () => {
      throw new Error(noTest + 'dragEnded()');
    });
  });
});
