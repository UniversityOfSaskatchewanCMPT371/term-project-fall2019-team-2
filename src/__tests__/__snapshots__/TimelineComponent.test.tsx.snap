// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimelineComponent /> R2 Tests
 T2.1 Snapshot Test
 1`] = `
"<TimelineComponent data={{...}}>
  <div>
    <div>
      <div>
        <button onClick={[Function: bound toggleTimeline]}>
          Switch to Interval Timeline
        </button>
        <label>
          Y-Axis
        </label>
        <select id=\\"ySelect\\" value=\\"Order ID\\" onChange={[Function: onChange]}>
          <option value=\\"Order ID\\">
            Order ID
          </option>
          <option value=\\"Units Sold\\">
            Units Sold
          </option>
          <option value=\\"Unit Price\\">
            Unit Price
          </option>
          <option value=\\"Unit Cost\\">
            Unit Cost
          </option>
          <option value=\\"Total Revenue\\">
            Total Revenue
          </option>
          <option value=\\"Total Cost\\">
            Total Cost
          </option>
          <option value=\\"Total Profit\\">
            Total Profit
          </option>
          <option value=\\"index\\">
            index
          </option>
          <option value=\\"Order Date_num\\">
            Order Date_num
          </option>
        </select>
        <label>
          Starting Range
        </label>
        <select id=\\"xSelect\\" value=\\"Order Date\\" onChange={[Function: onChange]}>
          <option value=\\"Order Date\\">
            Order Date
          </option>
          <option value=\\"Order ID\\">
            Order ID
          </option>
          <option value=\\"Ship Date\\">
            Ship Date
          </option>
          <option value=\\"Units Sold\\">
            Units Sold
          </option>
          <option value=\\"Unit Price\\">
            Unit Price
          </option>
          <option value=\\"Unit Cost\\">
            Unit Cost
          </option>
          <option value=\\"Total Revenue\\">
            Total Revenue
          </option>
          <option value=\\"Total Cost\\">
            Total Cost
          </option>
          <option value=\\"Total Profit\\">
            Total Profit
          </option>
          <option value=\\"index\\">
            index
          </option>
          <option value=\\"Order Date_num\\">
            Order Date_num
          </option>
        </select>
        <label>
          Ending Range
        </label>
        <select id=\\"x2Select\\" value=\\"Order Date\\" onChange={[Function: onChange]}>
          <option value=\\"Order Date\\">
            Order Date
          </option>
          <option value=\\"Order ID\\">
            Order ID
          </option>
          <option value=\\"Ship Date\\">
            Ship Date
          </option>
          <option value=\\"Units Sold\\">
            Units Sold
          </option>
          <option value=\\"Unit Price\\">
            Unit Price
          </option>
          <option value=\\"Unit Cost\\">
            Unit Cost
          </option>
          <option value=\\"Total Revenue\\">
            Total Revenue
          </option>
          <option value=\\"Total Cost\\">
            Total Cost
          </option>
          <option value=\\"Total Profit\\">
            Total Profit
          </option>
          <option value=\\"index\\">
            index
          </option>
          <option value=\\"Order Date_num\\">
            Order Date_num
          </option>
        </select>
      </div>
      <div id=\\"svgtarget\\" />
    </div>
  </div>
</TimelineComponent>"
`;

exports[`<TimelineComponent /> R2 Tests
 T2.1 Snapshot Test
 2`] = `
"<svg width=\\"934\\" height=\\"768\\">
<g transform=\\"translate(70, 0)\\">
<rect width=\\"934\\" height=\\"598\\" style=\\"fill: none;\\">
</rect>
<defs>
<clipPath id=\\"barsBox\\">
<rect width=\\"934\\" height=\\"768\\" x=\\"0\\" y=\\"0\\">
</rect>
</clipPath>
</defs>
<g clip-path=\\"url(#barsBox)\\">
<g id=\\"barsLayer\\">
<g class=\\"plot\\" id=\\"bars\\">
<rect class=\\"line pin\\" x=\\"0\\" width=\\"2\\" y=\\"419.4852605640208\\" height=\\"178.5147394359792\\">
</rect>
<rect class=\\"line pin\\" x=\\"48.5\\" width=\\"2\\" y=\\"45.45221988871617\\" height=\\"552.5477801112838\\">
</rect>
<rect class=\\"line pin\\" x=\\"97\\" width=\\"2\\" y=\\"0\\" height=\\"598\\">
</rect>
<rect class=\\"line pin\\" x=\\"145.5\\" width=\\"2\\" y=\\"598\\" height=\\"0\\">
</rect>
<rect class=\\"line pin\\" x=\\"194\\" width=\\"2\\" y=\\"221.82608227113172\\" height=\\"376.1739177288683\\">
</rect>
<circle cx=\\"0\\" cy=\\"419.4852605640208\\" r=\\"5\\" style=\\"fill: #69b3a2;\\" stroke=\\"black\\">
</circle>
<circle cx=\\"48.5\\" cy=\\"45.45221988871617\\" r=\\"5\\" style=\\"fill: #69b3a2;\\" stroke=\\"black\\">
</circle>
<circle cx=\\"97\\" cy=\\"0\\" r=\\"5\\" style=\\"fill: #69b3a2;\\" stroke=\\"black\\">
</circle>
<circle cx=\\"145.5\\" cy=\\"598\\" r=\\"5\\" style=\\"fill: #69b3a2;\\" stroke=\\"black\\">
</circle>
<circle cx=\\"194\\" cy=\\"221.82608227113172\\" r=\\"5\\" style=\\"fill: #69b3a2;\\" stroke=\\"black\\">
</circle>
<g class=\\"xtick\\" opacity=\\"1\\" transform=\\"translate(25,598)\\">
<line stroke=\\"blue\\" y2=\\"6\\">
</line>
<text style=\\"text-anchor: end; font-size: 1rem;\\" dx=\\"-.8em\\" dy=\\".15em\\" transform=\\"rotate(-90)\\">1/1/2010
</text>
</g>
</g>
</g>
</g>
<g id=\\"axisLayer\\">
<g style=\\"color: red;\\" class=\\"y axis\\" fill=\\"none\\" font-size=\\"10\\" font-family=\\"sans-serif\\" text-anchor=\\"end\\">
<path class=\\"domain\\" stroke=\\"currentColor\\" d=\\"M-6,598.5H0.5V0.5H-6\\">
</path>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,583.3875042617942)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">350,000,000
</text>
</g>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,531.0850872294376)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">400,000,000
</text>
</g>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,478.78267019708096)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">450,000,000
</text>
</g>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,426.48025316472433)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">500,000,000
</text>
</g>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,374.1778361323677)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">550,000,000
</text>
</g>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,321.87541910001113)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">600,000,000
</text>
</g>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,269.5730020676545)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">650,000,000
</text>
</g>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,217.27058503529787)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">700,000,000
</text>
</g>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,164.96816800294124)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">750,000,000
</text>
</g>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,112.66575097058455)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">800,000,000
</text>
</g>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,60.36333393822804)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">850,000,000
</text>
</g>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,8.060916905871295)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">900,000,000
</text>
</g>
<text transform=\\"rotate(-90)\\" y=\\"6\\" dy=\\".71em\\" style=\\"text-anchor: end; color: red;\\">yColumn
</text>
</g>
</g>
<text transform=\\"translate(467,618)\\" style=\\"text-anchor: start;\\">Order Date
</text>
<text transform=\\"rotate(-90)\\" y=\\"-70\\" x=\\"-299\\" dy=\\"1em\\" style=\\"text-anchor: middle;\\">Order ID
</text>
</g>
</svg>
"
`;

exports[`<TimelineComponent /> R2 Tests
 T2.4 Snapshot Test
 1`] = `
"<TimelineComponent data={{...}}>
  <div>
    <div>
      <div>
        <button onClick={[Function: bound toggleTimeline]}>
          Switch to Interval Timeline
        </button>
        <label>
          Y-Axis
        </label>
        <select id=\\"ySelect\\" value=\\"Order ID\\" onChange={[Function: onChange]}>
          <option value=\\"Order ID\\">
            Order ID
          </option>
          <option value=\\"Units Sold\\">
            Units Sold
          </option>
          <option value=\\"Unit Price\\">
            Unit Price
          </option>
          <option value=\\"Unit Cost\\">
            Unit Cost
          </option>
          <option value=\\"Total Revenue\\">
            Total Revenue
          </option>
          <option value=\\"Total Cost\\">
            Total Cost
          </option>
          <option value=\\"Total Profit\\">
            Total Profit
          </option>
          <option value=\\"index\\">
            index
          </option>
          <option value=\\"Order Date_num\\">
            Order Date_num
          </option>
        </select>
        <label>
          Starting Range
        </label>
        <select id=\\"xSelect\\" value=\\"Order Date\\" onChange={[Function: onChange]}>
          <option value=\\"Order Date\\">
            Order Date
          </option>
          <option value=\\"Order ID\\">
            Order ID
          </option>
          <option value=\\"Ship Date\\">
            Ship Date
          </option>
          <option value=\\"Units Sold\\">
            Units Sold
          </option>
          <option value=\\"Unit Price\\">
            Unit Price
          </option>
          <option value=\\"Unit Cost\\">
            Unit Cost
          </option>
          <option value=\\"Total Revenue\\">
            Total Revenue
          </option>
          <option value=\\"Total Cost\\">
            Total Cost
          </option>
          <option value=\\"Total Profit\\">
            Total Profit
          </option>
          <option value=\\"index\\">
            index
          </option>
          <option value=\\"Order Date_num\\">
            Order Date_num
          </option>
        </select>
        <label>
          Ending Range
        </label>
        <select id=\\"x2Select\\" value=\\"Order Date\\" onChange={[Function: onChange]}>
          <option value=\\"Order Date\\">
            Order Date
          </option>
          <option value=\\"Order ID\\">
            Order ID
          </option>
          <option value=\\"Ship Date\\">
            Ship Date
          </option>
          <option value=\\"Units Sold\\">
            Units Sold
          </option>
          <option value=\\"Unit Price\\">
            Unit Price
          </option>
          <option value=\\"Unit Cost\\">
            Unit Cost
          </option>
          <option value=\\"Total Revenue\\">
            Total Revenue
          </option>
          <option value=\\"Total Cost\\">
            Total Cost
          </option>
          <option value=\\"Total Profit\\">
            Total Profit
          </option>
          <option value=\\"index\\">
            index
          </option>
          <option value=\\"Order Date_num\\">
            Order Date_num
          </option>
        </select>
      </div>
      <div id=\\"svgtarget\\" />
    </div>
  </div>
</TimelineComponent>"
`;

exports[`<TimelineComponent /> R2 Tests
 T2.4 Snapshot Test
 2`] = `
"<svg width=\\"934\\" height=\\"768\\">
<g transform=\\"translate(70, 0)\\">
<rect width=\\"934\\" height=\\"598\\" style=\\"fill: none;\\">
</rect>
<defs>
<clipPath id=\\"barsBox\\">
<rect width=\\"934\\" height=\\"768\\" x=\\"0\\" y=\\"0\\">
</rect>
</clipPath>
</defs>
<g clip-path=\\"url(#barsBox)\\">
<g id=\\"barsLayer\\">
<g class=\\"plot\\" id=\\"bars\\">
<rect class=\\"bar\\" x=\\"125\\" width=\\"0\\" y=\\"419.4852605640208\\" height=\\"178.5147394359792\\" style=\\"fill: #61a3a9; opacity: 0.2;\\">
</rect>
<rect class=\\"bar\\" x=\\"125\\" width=\\"0\\" y=\\"45.45221988871617\\" height=\\"552.5477801112838\\" style=\\"fill: #61a3a9; opacity: 0.2;\\">
</rect>
<rect class=\\"bar\\" x=\\"125\\" width=\\"0\\" y=\\"0\\" height=\\"598\\" style=\\"fill: #61a3a9; opacity: 0.2;\\">
</rect>
<rect class=\\"bar\\" x=\\"125\\" width=\\"0\\" y=\\"598\\" height=\\"0\\" style=\\"fill: #61a3a9; opacity: 0.2;\\">
</rect>
<rect class=\\"bar\\" x=\\"125\\" width=\\"0\\" y=\\"221.82608227113172\\" height=\\"376.1739177288683\\" style=\\"fill: #61a3a9; opacity: 0.2;\\">
</rect>
<g class=\\"xtick\\" opacity=\\"1\\" transform=\\"translate(125 ,
                    598)\\">
<line stroke=\\"blue\\" y2=\\"6\\">
</line>
<text style=\\"text-anchor: end; font-size: 1rem;\\" dx=\\"-.8em\\" dy=\\".15em\\" transform=\\"rotate(-90)\\">1/1/2010
</text>
</g>
</g>
</g>
</g>
<g id=\\"axisLayer\\">
<g style=\\"color: red;\\" class=\\"y axis\\" fill=\\"none\\" font-size=\\"10\\" font-family=\\"sans-serif\\" text-anchor=\\"end\\">
<path class=\\"domain\\" stroke=\\"currentColor\\" d=\\"M-6,598.5H0.5V0.5H-6\\">
</path>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,583.3875042617942)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">350,000,000
</text>
</g>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,531.0850872294376)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">400,000,000
</text>
</g>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,478.78267019708096)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">450,000,000
</text>
</g>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,426.48025316472433)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">500,000,000
</text>
</g>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,374.1778361323677)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">550,000,000
</text>
</g>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,321.87541910001113)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">600,000,000
</text>
</g>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,269.5730020676545)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">650,000,000
</text>
</g>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,217.27058503529787)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">700,000,000
</text>
</g>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,164.96816800294124)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">750,000,000
</text>
</g>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,112.66575097058455)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">800,000,000
</text>
</g>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,60.36333393822804)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">850,000,000
</text>
</g>
<g class=\\"tick\\" opacity=\\"1\\" transform=\\"translate(0,8.060916905871295)\\">
<line stroke=\\"currentColor\\" x2=\\"-6\\">
</line>
<text fill=\\"currentColor\\" x=\\"-9\\" dy=\\"0.32em\\">900,000,000
</text>
</g>
<text transform=\\"rotate(-90)\\" y=\\"6\\" dy=\\".71em\\" style=\\"text-anchor: end; color: red;\\">yColumn
</text>
</g>
</g>
<text transform=\\"translate(477,618)\\" style=\\"text-anchor: start;\\">end: Order ID
</text>
<text transform=\\"translate(467,618)\\" style=\\"text-anchor: end;\\">start: Order Date,
</text>
<text transform=\\"rotate(-90)\\" y=\\"-70\\" x=\\"-299\\" dy=\\"1em\\" style=\\"text-anchor: middle;\\">Order ID
</text>
</g>
</svg>
"
`;
